mm/ktsan/stack_depot.c

```diff
@@ -0,0 +1,132 @@
+
+/* Only available during early boot. */
+void __init kt_stack_depot_init(kt_stack_depot_t *depot,
+			unsigned part_count, unsigned stack_count)
+{
+	unsigned i;
+
+	depot->part_count = part_count;
+	depot->stack_count = stack_count;
+
+	kt_cache_init(&depot->part_cache,
+			sizeof(*depot->parts) * part_count, 1);
+	depot->parts = kt_cache_alloc(&depot->part_cache);
+	BUG_ON(depot->parts == NULL);
+
+	for (i = 0; i < part_count; i++) {
+		depot->parts[i].head = NULL;
+		depot->parts[i].hash = 0;
+	}
+
+	kt_cache_init(&depot->stack_cache,
+		sizeof(kt_stack_depot_obj_t), stack_count);
+
+	kt_spin_init(&depot->lock);
+}
+
+/* Only available during early boot. */
+void __init kt_stack_depot_destroy(kt_stack_depot_t *depot)
+{
+	kt_cache_destroy(&depot->part_cache);
+	kt_cache_destroy(&depot->stack_cache);
+	depot->parts = NULL;
+}
+
+static inline u32 kt_stack_hash(kt_stack_t *stack)
+{
+	return jhash2((u32 *)stack->pc,
+			stack->size * sizeof(u32) / sizeof(stack->pc[0]),
+			KT_STACK_HASH_SEED);
+}
+
+static inline bool kt_stack_compare(kt_stack_t *first, kt_stack_t *second)
+{
+	if (first->size != second->size)
+		return false;
+	return memcmp(&first->pc, &second->pc,
+			sizeof(first->pc[0]) * first->size);
+}
+
+
+static inline kt_stack_depot_obj_t *kt_stack_depot_lookup_obj(
+	kt_stack_depot_t *depot, kt_stack_depot_obj_t *head, kt_stack_t *stack)
+{
+	kt_stack_depot_obj_t *obj;
+
+	for (obj = head; obj; obj = obj->link)
+		if (kt_stack_compare(&obj->stack, stack))
+			return obj;
+
+	return NULL;
+}
+
+static inline kt_stack_handle_t kt_stack_depot_obj_to_handle(
+			kt_stack_depot_t *depot, kt_stack_depot_obj_t *obj)
+{
+	return ((uptr_t)obj - depot->stack_cache.base) / sizeof(*obj);
+}
+
+static inline kt_stack_depot_obj_t *kt_stack_depot_handle_to_obj(
+			kt_stack_depot_t *depot, kt_stack_handle_t handle)
+{
+	return (kt_stack_depot_obj_t *)(depot->stack_cache.base +
+			handle * sizeof(kt_stack_depot_obj_t));
+}
+
+kt_stack_handle_t kt_stack_depot_save(kt_stack_depot_t *depot,
+					kt_stack_t *stack)
+{
+	u32 hash;
+	kt_stack_depot_part_t *part;
+	kt_stack_depot_obj_t *obj;
+
+	BUG_ON(stack->size == 0);
+
+	hash = kt_stack_hash(stack);
+	part = &depot->parts[hash % depot->part_count];
+
+	obj = kt_stack_depot_lookup_obj(depot,
+			smp_load_acquire(&part->head), stack);
+	if (obj)
+		return kt_stack_depot_obj_to_handle(depot, obj);
+
+	kt_spin_lock(&depot->lock);
+
+	obj = kt_stack_depot_lookup_obj(depot, part->head, stack);
+	if (!obj) {
+		kt_stack_depot_obj_t *new = kt_cache_alloc(&depot->stack_cache);
+		BUG_ON(new == NULL); /* Out of memory. */
+		new->stack = *stack;
+
+		new->link = obj;
+		smp_store_release(&part->head, new);
+
+		obj = new;
+	}
+
+	kt_spin_unlock(&depot->lock);
+
+	return kt_stack_depot_obj_to_handle(depot, obj);
+}
+
+kt_stack_handle_t kt_stack_depot_save_current(kt_stack_depot_t *depot,
+						unsigned long strip_addr)
+{
+	kt_stack_t stack;
+	kt_stack_save_current(&stack, strip_addr);
```

Comment:
kt_stack_save_current does not work any more (requires frame pointers)
kt_stack_depot_save_current is not used, so you can just delete it